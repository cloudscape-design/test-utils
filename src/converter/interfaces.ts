// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export type TestUtilType = 'dom' | 'selectors';

export interface ComponentMetadata {
  /**
   * Name of the component in pascal case.
   * Examples: Button, Alert, ButtonDropdown
   */
  name: string;

  /**
   * Plural name of the component in pascal case.
   * Examples: Buttons, Alerts, ButtonDropdowns
   */
  pluralName: string;

  /**
   * Folder name of the component test utils.
   * If not specified, the kebab case of the component name will be used by default.
   */
  testUtilsFolderName?: string;
}

export interface GenerateTestUtilsParams {
  /**
   * List of components metadata to generate test utils.
   */
  components: ComponentMetadata[];

  /*
   * Absolute path to the test utils folder.
   *
   * Component wrappers will be loaded from this path.
   * Generated test utils will be stored in this folder.
   *
   * Expected file name format is kebab case.
   *
   */
  testUtilsPath: string;
}

export interface ComponentWrapperMetadata extends ComponentMetadata {
  /*
   * Name of the component wrapper in pascal case
   * Examples: ButtonWrapper, AlertWrapper, ButtonDropdownWrapper
   */
  wrapperName: string;

  /**
   * Relative path to import the wrapper.
   */
  wrapperImportPath: string;
}

export interface ComponentPublicMetadata {
  /**
   * Plural name of the component in pascal case.
   * Examples: Buttons, Alerts, ButtonDropdowns
   */
  pluralName: string;

  /*
   * Name of the component wrapper in pascal case
   * Examples: ButtonWrapper, AlertWrapper, ButtonDropdownWrapper
   */
  wrapperName: string;
}
