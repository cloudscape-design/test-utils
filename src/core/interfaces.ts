// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export interface IElementWrapper<ElementType, CollectionType> {
  getElement(): ElementType;

  matches(selector: string): IElementWrapper<ElementType, CollectionType> | null;

  find(selector: string): IElementWrapper<ElementType, CollectionType> | null;

  findAll(selector: string): CollectionType;
}
