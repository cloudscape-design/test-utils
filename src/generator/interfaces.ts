// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

export type TestUtilType = 'dom' | 'selectors';

export interface ComponentMetadata {
  /**
   * Name of the component in pascal case.
   * Examples: Button, Alert, ButtonDropdown
   */
  name: string;

  /**
   * Plural name of the component in pascal case.
   * Examples: Buttons, Alerts, ButtonDropdowns
   */
  pluralName: string;

  /**
   * Test utils wrapper name of the component in pascal case.
   * Examples: ButtonWrapper, AlertWrapper, ButtonDropdownWrapper
   */
  wrapperName: string;

  /**
   * Path to import the test utils wrapper.
   */
  wrapperImportPath: string;
}

export type ComponentPublicMetadata = Pick<ComponentMetadata, 'pluralName' | 'wrapperName'>;
export type ComponentPublicMetadataMap = Record<string, ComponentPublicMetadata>;

export interface GenerateFindersParams {
  /**
   * List of the components that test utils will be generated for.
   */
  components: ComponentMetadata[];

  /**
   * Target test utils type. The values can be `dom` or `selectors`.
   */
  testUtilType: TestUtilType;

  /**
   * Extra imports to be added to the generated file head.
   */
  extraImports?: string[];
}
