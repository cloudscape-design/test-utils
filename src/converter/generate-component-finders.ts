// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { ComponentWrapperMetadata, TestUtilType } from './interfaces';

const componentWrapperImport = ({ wrapperName, wrapperImportPath }: ComponentWrapperMetadata) => `
import ${wrapperName} from '${wrapperImportPath}';`;

const componentWrapperExport = ({ wrapperName }: ComponentWrapperMetadata) => `
export { ${wrapperName} };`;

const componentFinders = ({ name, wrapperName, pluralName }: ComponentWrapperMetadata) => `
ElementWrapper.prototype.find${name} = function(selector) {
  const rootSelector = \`.$\{${wrapperName}.rootSelector}\`;
  // casting to 'any' is needed to avoid this issue with generics
  // https://github.com/microsoft/TypeScript/issues/29132
  return (this as any).findComponent(selector ? appendSelector(selector, rootSelector) : rootSelector, ${wrapperName});
};

ElementWrapper.prototype.findAll${pluralName} = function(selector) {
  return this.findAllComponents(${wrapperName}, selector);
};`;

const componentFindersInterfaces = {
  dom: ({ name, pluralName, wrapperName }: ComponentWrapperMetadata) => `
/**
 * Returns the wrapper of the first ${name} that matches the specified CSS selector.
 * If no CSS selector is specified, returns the wrapper of the first ${name}.
 * If no matching ${name} is found, returns \`null\`.
 *
 * @param {string} [selector] CSS Selector
 * @returns {${wrapperName} | null}
 */
find${name}(selector?: string): ${wrapperName} | null;

/**
 * Returns an array of ${name} wrapper that matches the specified CSS selector.
 * If no CSS selector is specified, returns all of the ${pluralName} inside the current wrapper.
 * If no matching ${name} is found, returns an empty array.
 *
 * @param {string} [selector] CSS Selector
 * @returns {Array<${wrapperName}>}
 */
findAll${pluralName}(selector?: string): Array<${wrapperName}>;`,

  selectors: ({ name, pluralName, wrapperName }: ComponentWrapperMetadata) => `
/**
 * Returns a wrapper that matches the ${pluralName} with the specified CSS selector.
 * If no CSS selector is specified, returns a wrapper that matches ${pluralName}.
 *
 * @param {string} [selector] CSS Selector
 * @returns {${wrapperName}}
 */
find${name}(selector?: string): ${wrapperName};

/**
 * Returns a multi-element wrapper that matches ${pluralName} with the specified CSS selector.
 * If no CSS selector is specified, returns a multi-element wrapper that matches ${pluralName}.
 *
 * @param {string} [selector] CSS Selector
 * @returns {MultiElementWrapper<${wrapperName}>}
 */
findAll${pluralName}(selector?: string): MultiElementWrapper<${wrapperName}>;`,
};

const defaultExport = {
  dom: `
export default function wrapper(root: Element = document.body) {
  if (document && document.body && !document.body.contains(root)) {
    console.warn('[AwsUi] [test-utils] provided element is not part of the document body, interactions may work incorrectly')
  };
  return new ElementWrapper(root);
}`,

  selectors: `
export default function wrapper(root: string = 'body') {
  return new ElementWrapper(root);
}`,
};

export interface GenerateFindersParams {
  components: ComponentWrapperMetadata[];
  testUtilType: TestUtilType;
}

export const generateComponentFinders = ({ components, testUtilType }: GenerateFindersParams) => `
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { ElementWrapper } from '@cloudscape-design/test-utils-core/${testUtilType}';
import { appendSelector } from '@cloudscape-design/test-utils-core/utils';

export { ElementWrapper };
${components.map(componentWrapperImport).join('')}

${components.map(componentWrapperExport).join('')}

declare module '@cloudscape-design/test-utils-core/dist/${testUtilType}' {
   interface ElementWrapper {
    ${components.map(componentFindersInterfaces[testUtilType]).join('')}
   }
}

${components.map(componentFinders).join('')}

${defaultExport[testUtilType]}
`;
