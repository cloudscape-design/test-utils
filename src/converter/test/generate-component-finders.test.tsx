// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { describe, test, expect } from 'vitest';
import { generateComponentFinders } from '../generate-component-finders';
import { ComponentWrapperMetadata } from '../interfaces';

const mockComponents: ComponentWrapperMetadata[] = [
  {
    name: 'Alert',
    pluralName: 'Alerts',
    wrapperName: 'AlertWrapper',
    wrapperImportPath: './test-utils/AlertWrapper',
    testUtilsFolderName: '../test-utils',
  },
  {
    name: 'Status',
    pluralName: 'Status', // The plural name is deliberately the same as singular
    wrapperName: 'StatusWrapper',
    wrapperImportPath: './test-utils/StatusWrapper',
    testUtilsFolderName: '../test-utils',
  },
];

describe(`${generateComponentFinders.name}`, () => {
  const testUtilTypes = ['dom', 'selectors'] as const;

  describe.each(testUtilTypes)('%s', testUtilType => {
    const sourceFileContent = generateComponentFinders({ components: mockComponents, testUtilType });

    test('it re-exports element wrapper', () => {
      expect(sourceFileContent).toMatch('export { ElementWrapper }');
    });

    test('it export component wrappers', () => {
      expect(sourceFileContent).toMatch('export { AlertWrapper };');
      expect(sourceFileContent).toMatch('export { StatusWrapper };');
    });

    test('it exports interfaces', () => {
      expect(sourceFileContent).toMatch(`declare module '@cloudscape-design/test-utils-core/dist/${testUtilType}'`);

      if (testUtilType === 'dom') {
        expect(sourceFileContent).toMatch(`findAlert(selector?: Selector): AlertWrapper | null`);
        expect(sourceFileContent).toMatch(`findAllAlerts(selector?: Selector): Array<AlertWrapper>`);
        expect(sourceFileContent).toMatch(`findStatus(selector?: Selector): StatusWrapper | null`);
        expect(sourceFileContent).toMatch(`findAllStatus(selector?: Selector): Array<StatusWrapper>`);
      } else {
        expect(sourceFileContent).toMatch(`findAlert(selector?: string): AlertWrapper`);
        expect(sourceFileContent).toMatch(`findAllAlerts(selector?: string): MultiElementWrapper<AlertWrapper>`);
        expect(sourceFileContent).toMatch(`findStatus(selector?: string): StatusWrapper`);
        expect(sourceFileContent).toMatch(`findAllStatus(selector?: string): MultiElementWrapper<StatusWrapper>`);
      }
    });

    test('it adds finder implementations to the ElementWrapper', () => {
      expect(sourceFileContent).toMatch('ElementWrapper.prototype.findAlert = function(selector)');
      expect(sourceFileContent).toMatch('ElementWrapper.prototype.findAllAlerts = function(selector)');
      expect(sourceFileContent).toMatch('ElementWrapper.prototype.findStatus = function(selector)');
      expect(sourceFileContent).toMatch('ElementWrapper.prototype.findAllStatus = function(selector)');
    });

    test('it exports the wrapper creator', () => {
      if (testUtilType === 'dom') {
        expect(sourceFileContent).toMatch('export default function wrapper(root: Element = document.body)');
      } else {
        expect(sourceFileContent).toMatch(`export default function wrapper(root: string = 'body')`);
      }
    });
  });
});
