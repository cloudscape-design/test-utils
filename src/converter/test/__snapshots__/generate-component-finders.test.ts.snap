// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates the finders for the specified list > generates component specific finders for dom 1`] = `
"
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { ElementWrapper } from '@cloudscape-design/test-utils-core/dom';
import { appendSelector } from '@cloudscape-design/test-utils-core/utils';

export { ElementWrapper };

import FooWrapper from 'foo/wrapper/import/path';
import BarWrapper from 'bar/wrapper/import/path';


export { FooWrapper };
export { BarWrapper };

declare module '@cloudscape-design/test-utils-core/dist/dom' {
   interface ElementWrapper {
    
/**
 * Returns the wrapper of the first Foo that matches the specified CSS selector.
 * If no CSS selector is specified, returns the wrapper of the first Foo.
 * If no matching Foo is found, returns \`null\`.
 *
 * @param {string} [selector] CSS Selector
 * @returns {FooWrapper | null}
 */
findFoo(selector?: string): FooWrapper | null;

/**
 * Returns an array of Foo wrapper that matches the specified CSS selector.
 * If no CSS selector is specified, returns all of the Foos inside the current wrapper.
 * If no matching Foo is found, returns an empty array.
 *
 * @param {string} [selector] CSS Selector
 * @returns {Array<FooWrapper>}
 */
findAllFoos(selector?: string): Array<FooWrapper>;
/**
 * Returns the wrapper of the first Bar that matches the specified CSS selector.
 * If no CSS selector is specified, returns the wrapper of the first Bar.
 * If no matching Bar is found, returns \`null\`.
 *
 * @param {string} [selector] CSS Selector
 * @returns {BarWrapper | null}
 */
findBar(selector?: string): BarWrapper | null;

/**
 * Returns an array of Bar wrapper that matches the specified CSS selector.
 * If no CSS selector is specified, returns all of the Bars inside the current wrapper.
 * If no matching Bar is found, returns an empty array.
 *
 * @param {string} [selector] CSS Selector
 * @returns {Array<BarWrapper>}
 */
findAllBars(selector?: string): Array<BarWrapper>;
   }
}


ElementWrapper.prototype.findFoo = function(selector) {
  const rootSelector = \`.\${FooWrapper.rootSelector}\`;
  // casting to 'any' is needed to avoid this issue with generics
  // https://github.com/microsoft/TypeScript/issues/29132
  return (this as any).findComponent(selector ? appendSelector(selector, rootSelector) : rootSelector, FooWrapper);
};

ElementWrapper.prototype.findAllFoos = function(selector) {
  return this.findAllComponents(FooWrapper, selector);
};
ElementWrapper.prototype.findBar = function(selector) {
  const rootSelector = \`.\${BarWrapper.rootSelector}\`;
  // casting to 'any' is needed to avoid this issue with generics
  // https://github.com/microsoft/TypeScript/issues/29132
  return (this as any).findComponent(selector ? appendSelector(selector, rootSelector) : rootSelector, BarWrapper);
};

ElementWrapper.prototype.findAllBars = function(selector) {
  return this.findAllComponents(BarWrapper, selector);
};


/**
 * Returns the component metadata including its plural and wrapper name.
 *
 * @param {string} componentName Component name in pascal case.
 * @returns {ComponentMetadata}
 */
export function getComponentMetadata(componentName: string) {
  return {"Foo":{"pluralName":"Foos","wrapperName":"FooWrapper"},"Bar":{"pluralName":"Bars","wrapperName":"BarWrapper"}}[componentName];
}


export default function wrapper(root: Element = document.body) {
  if (document && document.body && !document.body.contains(root)) {
    console.warn('[AwsUi] [test-utils] provided element is not part of the document body, interactions may work incorrectly')
  };
  return new ElementWrapper(root);
}
"
`;

exports[`generates the finders for the specified list > generates component specific finders for selectors 1`] = `
"
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { ElementWrapper } from '@cloudscape-design/test-utils-core/selectors';
import { appendSelector } from '@cloudscape-design/test-utils-core/utils';

export { ElementWrapper };

import FooWrapper from 'foo/wrapper/import/path';
import BarWrapper from 'bar/wrapper/import/path';


export { FooWrapper };
export { BarWrapper };

declare module '@cloudscape-design/test-utils-core/dist/selectors' {
   interface ElementWrapper {
    
/**
 * Returns a wrapper that matches the Foos with the specified CSS selector.
 * If no CSS selector is specified, returns a wrapper that matches Foos.
 *
 * @param {string} [selector] CSS Selector
 * @returns {FooWrapper}
 */
findFoo(selector?: string): FooWrapper;

/**
 * Returns a multi-element wrapper that matches Foos with the specified CSS selector.
 * If no CSS selector is specified, returns a multi-element wrapper that matches Foos.
 *
 * @param {string} [selector] CSS Selector
 * @returns {MultiElementWrapper<FooWrapper>}
 */
findAllFoos(selector?: string): MultiElementWrapper<FooWrapper>;
/**
 * Returns a wrapper that matches the Bars with the specified CSS selector.
 * If no CSS selector is specified, returns a wrapper that matches Bars.
 *
 * @param {string} [selector] CSS Selector
 * @returns {BarWrapper}
 */
findBar(selector?: string): BarWrapper;

/**
 * Returns a multi-element wrapper that matches Bars with the specified CSS selector.
 * If no CSS selector is specified, returns a multi-element wrapper that matches Bars.
 *
 * @param {string} [selector] CSS Selector
 * @returns {MultiElementWrapper<BarWrapper>}
 */
findAllBars(selector?: string): MultiElementWrapper<BarWrapper>;
   }
}


ElementWrapper.prototype.findFoo = function(selector) {
  const rootSelector = \`.\${FooWrapper.rootSelector}\`;
  // casting to 'any' is needed to avoid this issue with generics
  // https://github.com/microsoft/TypeScript/issues/29132
  return (this as any).findComponent(selector ? appendSelector(selector, rootSelector) : rootSelector, FooWrapper);
};

ElementWrapper.prototype.findAllFoos = function(selector) {
  return this.findAllComponents(FooWrapper, selector);
};
ElementWrapper.prototype.findBar = function(selector) {
  const rootSelector = \`.\${BarWrapper.rootSelector}\`;
  // casting to 'any' is needed to avoid this issue with generics
  // https://github.com/microsoft/TypeScript/issues/29132
  return (this as any).findComponent(selector ? appendSelector(selector, rootSelector) : rootSelector, BarWrapper);
};

ElementWrapper.prototype.findAllBars = function(selector) {
  return this.findAllComponents(BarWrapper, selector);
};


/**
 * Returns the component metadata including its plural and wrapper name.
 *
 * @param {string} componentName Component name in pascal case.
 * @returns {ComponentMetadata}
 */
export function getComponentMetadata(componentName: string) {
  return {"Foo":{"pluralName":"Foos","wrapperName":"FooWrapper"},"Bar":{"pluralName":"Bars","wrapperName":"BarWrapper"}}[componentName];
}


export default function wrapper(root: string = 'body') {
  return new ElementWrapper(root);
}
"
`;
